components/typography/Base/Ellipsis.tsx

```diff
@@ -0,0 +1,211 @@
+  enabledMeasure?: boolean;
+  text?: React.ReactNode;
+  width: number;
+  rows: number;
+  children: (cutChildren: React.ReactNode[], needEllipsis: boolean) => React.ReactNode;
+  onEllipsis: (isEllipsis: boolean) => void;
+}
+
+function cuttable(node: React.ReactElement) {
+  const type = typeof node;
+  return type === 'string' || type === 'number';
+}
+
+function getNodesLen(nodeList: React.ReactElement[]) {
+  let totalLen = 0;
+
+  nodeList.forEach(node => {
+    if (cuttable(node)) {
+      totalLen += String(node).length;
+    } else {
+      totalLen += 1;
+    }
+  });
+
+  return totalLen;
+}
+
+function sliceNodes(nodeList: React.ReactElement[], len: number) {
+  let currLen = 0;
+  const currentNodeList: React.ReactNode[] = [];
+
+  for (let i = 0; i < nodeList.length; i += 1) {
+    // Match to return
+    if (currLen === len) {
+      return currentNodeList;
+    }
+
+    const node = nodeList[i];
+    const canCut = cuttable(node);
+    const nodeLen = canCut ? String(node).length : 1;
+    const nextLen = currLen + nodeLen;
+
+    // Exceed but current not which means we need cut this
+    // This will not happen on validate ReactElement
+    if (nextLen > len) {
+      const restLen = len - currLen;
+      currentNodeList.push(String(node).slice(0, restLen));
+      return currentNodeList;
+    }
+
+    currentNodeList.push(node);
+    currLen = nextLen;
+  }
+
+  return nodeList;
+}
+
+const NONE = 0;
+const PREPARE = 1;
+const WALKING = 2;
+const DONE_WITH_ELLIPSIS = 3;
+const DONE_WITHOUT_ELLIPSIS = 4;
+
+const Ellipsis = ({ enabledMeasure, children, text, width, rows, onEllipsis }: EllipsisProps) => {
+  const [cutLength, setCutLength] = React.useState<[number, number, number]>([0, 0, 0]);
+  const [walkingState, setWalkingState] = React.useState<
+    | typeof NONE
+    | typeof PREPARE
+    | typeof WALKING
+    | typeof DONE_WITH_ELLIPSIS
+    | typeof DONE_WITHOUT_ELLIPSIS
+  >(NONE);
+  const [startLen, midLen, endLen] = cutLength;
+
+  const [singleRowHeight, setSingleRowHeight] = React.useState(0);
+
+  const singleRowRef = React.useRef<HTMLSpanElement>(null);
+  const midRowRef = React.useRef<HTMLSpanElement>(null);
+
+  const nodeList = React.useMemo(() => toArray(text), [text]);
+  const totalLen = React.useMemo(() => getNodesLen(nodeList), [nodeList]);
+
+  const mergedChildren = React.useMemo(() => {
+    if (!enabledMeasure || walkingState !== DONE_WITH_ELLIPSIS) {
+      return children(nodeList, false);
+    }
+
+    return children(sliceNodes(nodeList, midLen), midLen < totalLen);
+  }, [enabledMeasure, walkingState, children, nodeList, midLen, totalLen]);
+
+  // ======================== Walk ========================
+  React.useLayoutEffect(() => {
+    if (enabledMeasure && width && totalLen) {
+      setWalkingState(PREPARE);
+      setCutLength([0, Math.ceil(totalLen / 2), totalLen]);
+    }
+  }, [enabledMeasure, width, text, totalLen, rows]);
+
+  React.useLayoutEffect(() => {
```

Comment:
useLayoutEffect generates the following warning in the SSR environment.

Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format.

Can you consider the SSR environment without directly using useLayoutEffect in the antd and react-component packages?

Reference:
https://github.com/andarist/use-isomorphic-layout-effect/blob/master/src/index.ts

Reply:
@zombieJ

Reply:
I see the change but somehow React is still complaining about the useLayoutEffect at `line77` on file: 
`webpack-internal:///./node_modules/antd/lib/typography/Base/Ellipsis.js`